RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETMemberMetricCalculator.cs

```diff
@@ -121,5 +137,404 @@ private int CountOccurrences(string text, string pattern)
+        }
+
+        private int CountBinaryExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.AddExpression) ||
+                            n.IsKind(SyntaxKind.SubtractExpression) ||
+                            n.IsKind(SyntaxKind.MultiplyExpression) ||
+                            n.IsKind(SyntaxKind.DivideExpression) ||
+                            n.IsKind(SyntaxKind.ModuloExpression) ||
+                            n.IsKind(SyntaxKind.LeftShiftExpression) ||
+                            n.IsKind(SyntaxKind.RightShiftExpression));
+        }
+
+        private int CountUnaryExpressions(MemberDeclarationSyntax method)
+        {
+            int count = method.DescendantNodes().OfType<PrefixUnaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.PreIncrementExpression) ||
+                            n.IsKind(SyntaxKind.PreDecrementExpression) ||
+                            n.IsKind(SyntaxKind.UnaryPlusExpression) ||
+                            n.IsKind(SyntaxKind.UnaryMinusExpression));
+            count += method.DescendantNodes().OfType<PostfixUnaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.PostIncrementExpression) ||
+                            n.IsKind(SyntaxKind.PostDecrementExpression));
+            return count;
+        }
+
+        private int CountNumberOfParenthesizedExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<ExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.ParenthesizedExpression));
+        }
+
+        private int CountNumberOfLambdaExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LambdaExpressionSyntax>().Count();
+        }
+
+        private int CountMaxNestedBlocks(MemberDeclarationSyntax method)
+        {
+            if (!method.Kind().Equals(SyntaxKind.MethodDeclaration))
+            {
+                return 0;
+            }
+            BaseMethodDeclarationSyntax baseMethod = (BaseMethodDeclarationSyntax)method;
+            var blocks = baseMethod.Body.DescendantNodes().OfType<BlockSyntax>().ToList();
+
+            List<int> blocksAncestors = new List<int>();
+            foreach (var block in blocks)
+            {
+                blocksAncestors.Add(block.Ancestors().Count(a => a.IsKind(SyntaxKind.Block)));
+            }
+
+            if (!blocksAncestors.Any())
+            {
+                return 0;
+            }
+            return blocksAncestors.Max();
+        }
+
+        private int CountNumberOfUniqueWords(MemberDeclarationSyntax method)
+        {
+            if (!method.Kind().Equals(SyntaxKind.MethodDeclaration))
+            {
+                return 0;
+            }
+            BaseMethodDeclarationSyntax baseMethod = (BaseMethodDeclarationSyntax)method;
```

Comment:
Do you need code from 272 to 277? Look how the LOC method is used - member.ToString().

Reply:
I wanted to count the words in the method body, so it was necessary to do the cast. As for the first check (whether a member is actually a method) - when all the tests are run, at some point a non-method parameter is passed, and the cast fails.